### PR TITLE
Add method to retrieve the namespace names for the Jdbc Admin

### PR DESCRIPTION
**Pull request by Torch3333**
This adds the `admin.getNamespaceNames()` method which returns a set of all the namespaces names created through Scalar DB. To manage the created namespace names, a new table called `namespace` is added to the same namespace as the metadata table. 
This PR contains only the implementation for the JDBC Admin, the other Admin implementations will be added in following PRs.

Besides, we will also add a new method to the admin called `upgrade()` so that users can update their existing environment to be compatible with the new Scalar DB version using the `namespace` table. This will be added in a following PR.

### CODE REVIEW

File: core/src/main/java/com/scalar/db/storage/jdbc/JdbcConfig.java [hunk 1]
```diff
@@ -22,8 +25,10 @@ public class JdbcConfig {
       PREFIX + "prepared_statements_pool.max_open";
 
   public static final String ISOLATION_LEVEL = PREFIX + "isolation_level";
+  /** @deprecated As of 5.0, will be removed. Use {@link #METADATA_SCHEMA} instead. */
+  @Deprecated public static final String TABLE_METADATA_SCHEMA = PREFIX + "table_metadata.schema";
 
-  public static final String TABLE_METADATA_SCHEMA = PREFIX + "table_metadata.schema";
+  public static final String METADATA_SCHEMA = PREFIX + "metadata.schema";
```
Comment (Torch3333) [inline core/src/main/java/com/scalar/db/storage/jdbc/JdbcConfig.java:31]: The config `table_metadata.schema` is renamed to `metadata.schema` as this schema now contains the metadata table and namespace table.

File: core/src/main/java/com/scalar/db/storage/jdbc/JdbcAdmin.java [hunk 2]
```diff
@@ -162,6 +164,8 @@ public void createNamespace(String namespace, Map<String, String> options)
       } else {
         execute(connection, "CREATE SCHEMA " + fullNamespace);
       }
+      createNamespaceTableIfNotExists(connection);
+      insertIntoNamespaceTable(connection, namespace);
```
Comment (Torch3333) [inline core/src/main/java/com/scalar/db/storage/jdbc/JdbcAdmin.java:168]: When creating a namespace, the namespace name is added to the `namespace` table.

File: core/src/main/java/com/scalar/db/storage/jdbc/JdbcAdmin.java [hunk 3]
```diff
@@ -1049,4 +1058,89 @@ private String enclose(String name) {
...
+      // query
+      if ((rdbEngine == RdbEngine.MYSQL && (e.getErrorCode() == 1049 || e.getErrorCode() == 1146))
+          || (rdbEngine == RdbEngine.POSTGRESQL && "42P01".equals(e.getSQLState()))
+          || (rdbEngine == RdbEngine.ORACLE && e.getErrorCode() == 942)
+          || (rdbEngine == RdbEngine.SQL_SERVER && e.getErrorCode() == 208)) {
+        return Collections.emptySet();
+      }
+      throw new ExecutionException("getting the namespace names failed", e);
+    }
+  }
```
Comment (Torch3333) [inline core/src/main/java/com/scalar/db/storage/jdbc/JdbcAdmin.java:1085]: The `JdbcAdmin.getNamespaceNames()` implementation.

File: core/src/main/java/com/scalar/db/storage/jdbc/JdbcAdmin.java [hunk 4]
```diff
@@ -565,6 +572,8 @@ public void dropNamespace(String namespace) throws ExecutionException {
 
     try (Connection connection = dataSource.getConnection()) {
       execute(connection, dropStatement);
+      deleteFromNamespaceTable(connection, namespace);
+      deleteNamespaceTableIfEmpty(connection);
```
Comment (Torch3333) [inline core/src/main/java/com/scalar/db/storage/jdbc/JdbcAdmin.java:576]: When dropping a namespace, the namespace is also deleted from the `namespace` table.

File: core/src/main/java/com/scalar/db/storage/rpc/GrpcAdmin.java [hunk 5]
```diff
@@ -296,17 +298,14 @@ public void repairTable(
...
-                                .putAllOptions(options)
-                                .build())));
+            stub.withDeadlineAfter(config.getDeadlineDurationMillis(), TimeUnit.MILLISECONDS)
+                .repairTable(
+                    RepairTableRequest.newBuilder()
+                        .setNamespace(namespace)
+                        .setTable(table)
+                        .setTableMetadata(ProtoUtils.toTableMetadata(metadata))
+                        .putAllOptions(options)
+                        .build()));
```
Comment (Torch3333) [inline core/src/main/java/com/scalar/db/storage/rpc/GrpcAdmin.java:308]: The `execute()` method was nested twice.

File: core/src/main/java/com/scalar/db/storage/rpc/GrpcAdmin.java [hunk 6]
```diff
@@ -315,17 +314,25 @@ public void addNewColumnToTable(
...
-                                .setColumnType(ProtoUtils.toDataType(columnType))
-                                .build())));
+            stub.withDeadlineAfter(config.getDeadlineDurationMillis(), TimeUnit.MILLISECONDS)
+                .addNewColumnToTable(
+                    AddNewColumnToTableRequest.newBuilder()
+                        .setNamespace(namespace)
+                        .setTable(table)
+                        .setColumnName(columnName)
+                        .setColumnType(ProtoUtils.toDataType(columnType))
+                        .build()));
```
Comment (Torch3333) [inline core/src/main/java/com/scalar/db/storage/rpc/GrpcAdmin.java:324]: Here too.

File: core/src/main/java/com/scalar/db/storage/multistorage/MultiStorageAdmin.java [hunk 7]
```diff
@@ -182,6 +184,16 @@ public void addNewColumnToTable(
...
 
+  @Override
+  public Set<String> getNamespaceNames() throws ExecutionException {
+    Set<String> namespaceNames = new HashSet<>();
+    for (DistributedStorageAdmin admin : admins) {
+      namespaceNames.addAll(admin.getNamespaceNames());
+    }
+
+    return namespaceNames;
+  }
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/storage/multistorage/MultiStorageAdmin.java:195]: It looks like this method returns namespaces that are not in the namespace mapping. I'm not sure but is that okay?

**Review by brfrn169 — COMMENTED** [review]
Thank you! LGTM overall. Left several minor comments. Please take a look when you have time!

File: core/src/main/java/com/scalar/db/storage/multistorage/MultiStorageAdmin.java (hunk 7 above)
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/storage/multistorage/MultiStorageAdmin.java:195]: I see. Thank you for pointing this out. 
The configuration `multi-storage.table_mapping`  was to address the case where a given namespace would contain Scalar DB table and other non-Scalar DB tables? 
In that case, returning namespaces only from the `multi_storage.namespace_mapping`config is probably more appropriate.  
Concerning the `multi_storage.default_storage` admin, do you think we should return its namespaces ?

File: core/src/main/java/com/scalar/db/storage/jdbc/JdbcAdmin.java [hunk 8]
```diff
@@ -49,6 +49,8 @@ public class JdbcAdmin implements DistributedStorageAdmin {
   @VisibleForTesting static final String METADATA_COL_CLUSTERING_ORDER = "clustering_order";
   @VisibleForTesting static final String METADATA_COL_INDEXED = "indexed";
   @VisibleForTesting static final String METADATA_COL_ORDINAL_POSITION = "ordinal_position";
+  private static final String NAMESPACE_TABLE = "namespace";
```
[the remark this answers is not in the document]
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/storage/jdbc/JdbcAdmin.java:52]: It seems the case. Thank you. I will rename it.

File: core/src/main/java/com/scalar/db/storage/multistorage/MultiStorageAdmin.java (hunk 7 above)
Comment (brfrn169) [reply core/src/main/java/com/scalar/db/storage/multistorage/MultiStorageAdmin.java:195]: `multi-storage.table_mapping` is a deprecated config, so let's ignore it here.

Maybe returning namespaces from the `multi_storage.namespace_mappingconfig` config and the `multi_storage.default_storage` admin is more appropriate?

I think, for example, if users get namespaces from `getNamespaceNames()` but they face a `namespace not found` error when they call `getNamespaceTableNames()` for the namespace, it's very confusing behavior. This happens when `getNamespaceNames()` returns namespaces that are neither in the mapping nor the `multi_storage.default_storage` admin.

From this point of view, I feel that returning namespaces from the `multi_storage.namespace_mappingconfig` config and the `multi_storage.default_storage` admin is more appropriate.

What do you think?

Comment (Torch3333) [reply core/src/main/java/com/scalar/db/storage/multistorage/MultiStorageAdmin.java:195]: > From this point of view, I feel that returning namespaces from the multi_storage.namespace_mappingconfig config and the multi_storage.default_storage admin is more appropriate.

I agree. I will revise it accordingly.

**Comment by Torch3333** [comment]
@brfrn169 I revised the PR according to your comments. Please have a look again.

**Review by brfrn169 — COMMENTED** [review]
Left a comment. Other than that LGTM. Please take a look when you have time!

File: core/src/main/java/com/scalar/db/storage/multistorage/MultiStorageAdmin.java [hunk 9]
```diff
@@ -182,6 +183,19 @@ public void addNewColumnToTable(
...
+    Set<DistributedStorageAdmin> admins = new HashSet<>(namespaceAdminMap.values());
+    admins.add(defaultAdmin);
+
+    Set<String> namespaceNames = new HashSet<>();
+    for (DistributedStorageAdmin admin : admins) {
+      namespaceNames.addAll(admin.getNamespaceNames());
+    }
+
+    return namespaceNames;
+  }
```
[the remark this answers is not in the document]
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/storage/multistorage/MultiStorageAdmin.java]: Thank you for checking thoroughly, I did not understand it correctly.
I revised the implementation, please have a look again. 🙇

**Review by brfrn169 — APPROVED** [review]
LGTM! Thank you!

**Review by feeblefakie — APPROVED** [review]
LGTM! Thank you!

**Pull request merged**